### ISSUE TITLE
Bug: [no-unnecessary-condition] incorrect optional chaining handling for string type keys

### ISSUE THREAD

**Issue by mgiedrius** [typescript-eslint/typescript-eslint#5225]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.6.2&sourceType=module&code=PQKgBApgzgNglgOwC5gEQAEkE8AO0DGATnDkgLTTzLAID2ZArgghPtFAIaFZn60IATOEjj9UALjABtVBEKFahVAF0wIYACgN2PGAAiHJBzABeMAG8NYaQGsIWSVCTEEAc2WTL16wDNatR2dEVyswAF8NCI0+BCcwAUMOSQMjUwswgG5o-ji7BzAnF1c0gHI8kqyNYGAAUQAPPHwRNzAkAAs4KFbaMAAjCDAANw54AWzY2hgIADoYWlcACgSjKTzlAH5pv1oASgygA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A

### Repro Code

```typescript
/* eslint "@typescript-eslint/no-unnecessary-condition": ["error"] */

type Data = {
  [key: string]: {
    foo: string
  }
}

const data: Data = {};
const key: string = 'key';

//Expecting this to be valid
console.log(data[key]?.foo);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {}
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

Expecting `console.log(data[key]?.foo);` not to throw `Unnecessary optional chain on a non-nullish value.` error.

### Actual Result

`console.log(data[key]?.foo);` throws `Unnecessary optional chain on a non-nullish value.` error.

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.29.0` |
| `@typescript-eslint/parser` | `5.29.0` |
| `TypeScript` | `4.6.2` |
| `ESLint` | `8.15.0` |
| `node` | `web` |

**Comment by bradzacher** [comment]
This is correct.

The type of `dat[key]` is an object. That type is not nullish, thus the rule reports correctly.

If you want the rule to not error here the. Either manually add `undefined` to your type, or use [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess)

**Comment by mgiedrius** [comment]
@bradzacher Could you give more details on why is this valid? I'm a bit confused. The example I added when running in the browser throws `Uncaught TypeError: Cannot read properties of undefined (reading 'status')` when used without optional chaining.

```javascript
const data = {}
const key = 'key'
console.log(data[key].status)
```

**Comment by bradzacher** [comment]
Because

> The type of data[key] is an object. That type is not nullish

The rule trusts the types are correct. If your types tell us something is non nullish - then we have no choice to believe it.

I gave you options for correcting the type. I strongly suggest the compiler option.
